### PR TITLE
Web map tile service imagery provider

### PR DESCRIPTION
**Pull request by adamdavidcole**
Created function createWebTileServiceImageryProvider which utilizes the generic UrlTemplateImageryProvider
@pjcozzi @tiffanylu

### CODE REVIEW

**Comment by adamdavidcole** [comment]
I get strange behavior when I use this imagery provider. For instance in http://localhost:8080/Apps/Sandcastle/index.html?src=Imagery%20Layers%20Manipulation.html&label=Beginner changing the dropdown to USGS and unchecking all the options except tile coordinates and then zooming past 6 levels causes problems. 
However, I get the same results with the current implementation of Web map tile service imagery provider (which can be seen by following the same steps here: http://cesiumjs.org/Cesium/Apps/Sandcastle/index.html?src=Imagery%20Layers%20Manipulation.html&label=Beginner) 

Let me know if you have an idea why this is the case.

Still to do in this pull request is to finish the deprecation process for Web map tile service imagery provider. 

Thanks!
Adam

**Comment by pjcozzi** [comment]
> Let me know if you have an idea why this is the case.

This _could_ just be an issue with the data the service is returning, not the request Cesium is making.  I am not 100% sure though.  Can you please submit a separate issue about it?

File: .idea/misc.xml [hunk 1]
```diff
@@ -1,13 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (pjcozzi) [inline .idea/misc.xml]: Did you mean to remove this?

**Comment by pjcozzi** [comment]
> Still to do in this pull request is to finish the deprecation process for Web map tile service imagery provider.

Put a [task list](https://github.com/blog/1375%0A-task-lists-in-gfm-issues-pulls-comments) in the top comment of this pull request to manage what is left like in #3225 for example.

**Comment by pjcozzi** [comment]
Part of #2814.

@kring in #2814, you said implementing `WebMapTileServiceImageryProvider` with `UrlTemplateImageryProvider` _may_ be complicated.  This looks straightforward.  Are we missing something?

**Comment by pjcozzi** [comment]
Did you run the tests?  There are several failures:

image

Never open a pull request with failing tests unless part of the pull request comment is asking for help on why they are failing.

**Comment by adamdavidcole** [comment]
In my recent update, I fixed the way the urlTemplate is constructed and completed the deprecation process for WebMapTileServiceImageryProvider. I listed it for deprecation in 1.19, does that sound right? When tested by hand, the map imagery shows up as expected (aka. the same way it shows up on cesiumjs.org currently). 

After these changes I'm still failing 5 tests  in the spec, mostly the result of two separate problems. The first is that two tests expect to access the format by using 'provider.format' (seen here: https://github.com/tiffanylu/cesium/blob/webMapTileServiceImageryProvider/Specs/Scene/createWebMapTileServiceImageryProviderSpec.js#L223 and https://github.com/tiffanylu/cesium/blob/webMapTileServiceImageryProvider/Specs/Scene/createWebMapTileServiceImageryProviderSpec.js#L254). I believe that this issue should be resolved by removing those two line from the spec code.

Another two tests fail because WebMapTileServiceImageryProvider provides an option for passing tileMatrixLabels which map labels to strings. As a result tests like this fail: 
Expected '1' to equal 'second'." As of now I do not think there is any way to set up labels in UrlTemplateImageryProvider. One option here is to remove providing labels as an option for createWebTileServiceImageryProvider. 

Finally, the last test which fails is this: Expected function 'getTileCredits' to exist on WebMapTileServiceImageryProvider because it should implement interface ImageryProvider. I can't figure out why this fails because UrlTemplateImageryProvider does define the function getTileCredits: https://github.com/tiffanylu/cesium/blob/webMapTileServiceImageryProvider/Source/Scene/UrlTemplateImageryProvider.js#L447.

Please let me know what you think I should do to solve these testing issues. 

Thanks!

**Comment by pjcozzi** [comment]
@adamdavidcole can you merge in master.  There is probably a trivial conflict in CHANGES.md.

**Comment by pjcozzi** [comment]
> I listed it for deprecation in 1.19, does that sound right? 

Yes

**Comment by pjcozzi** [comment]
> The first is that two tests expect to access the format by using 'provider.format' (seen here: https://github.com/tiffanylu/cesium/blob/webMapTileServiceImageryProvider/Specs/Scene/createWebMapTileServiceImageryProviderSpec.js#L223 and https://github.com/tiffanylu/cesium/blob/webMapTileServiceImageryProvider/Specs/Scene/createWebMapTileServiceImageryProviderSpec.js#L254). I believe that this issue should be resolved by removing those two line from the spec code.

This is fine since it is a breaking change; `createWebMapTileServiceImageryProvider` returns a `UrlTemplateImageProvider`, which does not have this property.

However, please make sure that the KVP requests that use the `format` still work, e.g., https://github.com/tiffanylu/cesium/blob/webMapTileServiceImageryProvider/Source/Scene/WebMapTileServiceImageryProvider.js#L201

**Comment by pjcozzi** [comment]
> Another two tests fail because WebMapTileServiceImageryProvider provides an option for passing tileMatrixLabels which map labels to strings. As a result tests like this fail: 
Expected '1' to equal 'second'." As of now I do not think there is any way to set up labels in UrlTemplateImageryProvider. One option here is to remove providing labels as an option for createWebTileServiceImageryProvider.

I think you need to add support to `UrlTemplateImageryProvider` for this in one form or another, e.g., see https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Scene/WebMapTileServiceImageryProvider.js#L167

**Comment by pjcozzi** [comment]
> Finally, the last test which fails is this: Expected function 'getTileCredits' to exist on WebMapTileServiceImageryProvider because it should implement interface ImageryProvider. I can't figure out why this fails because UrlTemplateImageryProvider does define the function getTileCredits: https://github.com/tiffanylu/cesium/blob/webMapTileServiceImageryProvider/Source/Scene/UrlTemplateImageryProvider.js#L447.

Replace this test with a test like this: https://github.com/tiffanylu/cesium/blob/webMapTileServiceImageryProvider/Specs/Scene/createOpenStreetMapImageryProviderSpec.js#L34

**Comment by pjcozzi** [comment]
@adamdavidcole thanks for your initial work on this; we're going to hold off for now.  Hope you are having a great summer and thanks again for all your past contributions to Cesium!

**Pull request closed without merging**